Reliable releases distribution for tree-sitter CLI and other components

From https://github.com/tree-sitter/tree-sitter/issues/930#issuecomment-801519577:

> It would be awesome to automate release process for all official tree-sitter tools, especially for tree-sitter-cli, for all official bindings Wasm, Rust, Node.js, Python, Haskell, Ruby and the Playground with its separately living parsers and keep all in sync with the core tree-sitter library releases. This would help to reduce misunderstanding and situations that some things work somewhere and somewhere don't.
> - For now **tree-sitter-cli** installation from the [crate](https://crates.io/crates/tree-sitter-cli) seems the bad idea, the crate is stuck in 2 years old version.
> - #1122 - _tree-sitter-highlight 0.19.2 does not compile with tree-sitter 0.19.5_ - demonstrates an issue that changes in tree-sitter's Rust binding requires bumping version in all dependencies that use changed parts. Otherwise there need to be a CI check that would test that the last dependent can be built against all equal or higher versions of the dependence.
> - I can't say about all bindings but Node and Python bindings use **static** linking to tree-sitter core library and this means that these are lag behind the core library and don't receiving core fixes and logic improvements synchronously. IMO that's the important reason why such updates need to be automated. This doesn't solves problem with the core lib features covering but at least bug fixes would be delivered in time.

_I've previously posted a set of actual live badges there https://github.com/tree-sitter/tree-sitter/issues/930#issuecomment-801519577_
_But it makes sense to duplicate these here in this issue (all badges are clickable)._
### Versions
[![NPM](http://img.shields.io/npm/v/tree-sitter-cli.svg?label=CLI%20-%20npm&color=%23BF4A4A)](https://www.npmjs.org/package/tree-sitter-cli) [![crates.io](https://img.shields.io/crates/v/tree-sitter-cli.svg?label=CLI%20-%20crate&color=%23B48723)](https://crates.io/crates/tree-sitter-cli)

[![crates.io](https://img.shields.io/crates/v/tree-sitter-loader.svg?label=LOADER%20-%20crate&color=%23B48723)](https://crates.io/crates/tree-sitter-loader) [![crates.io](https://img.shields.io/crates/v/tree-sitter-config.svg?label=CONFIG%20-%20crate&color=%23B48723)](https://crates.io/crates/tree-sitter-config)

[![crates.io](https://img.shields.io/crates/v/tree-sitter-highlight.svg?label=HIGHLIGHT%20-%20crate&color=%23B48723)](https://crates.io/crates/tree-sitter-highlight) [![crates.io](https://img.shields.io/crates/v/tree-sitter-tags.svg?label=TAGS%20-%20crate&color=%23B48723)](https://crates.io/crates/tree-sitter-tags)


#### Bindings
[![binding:wasm:npm](http://img.shields.io/npm/v/web-tree-sitter.svg?label=WASM%20-%20npm&color=%23BF4A4A)](https://www.npmjs.org/package/web-tree-sitter) [![binding:rust](https://img.shields.io/crates/v/tree-sitter.svg?label=Rust&color=%23B48723)](https://crates.io/crates/tree-sitter) [![binding:node](http://img.shields.io/npm/v/tree-sitter.svg?label=Node&color=%23BF4A4A)](https://www.npmjs.org/package/tree-sitter)

[![binding:python](http://img.shields.io/pypi/v/tree-sitter.svg?label=Python&color=%233B6DAB)](https://pypi.org/project/tree-sitter) [![binding:haskell](http://img.shields.io/hackage/v/tree-sitter.svg?label=Haskell&color=%235A5182)](http://hackage.haskell.org/package/tree-sitter) [![binding:ruby](http://img.shields.io/gem/v/tree-sitter.svg?label=Ruby&color=%23CE3E2D)](https://rubygems.org/gems/tree-sitter)

The [0.20.0](https://github.com/tree-sitter/tree-sitter/releases/tag/v0.20.0) release happened 2 days ago and on above badges it's seen that it doesn't exist in **any** distribution channel except directly download from Github artifacts, use OS package managers (if these have it) or build from the source. CLI on the [crates.io](https://crates.io/crates/tree-sitter-cli) stuck in the past for 2 years.
As I understand current release process, there are a lot of manual actions that need to be happen.

Also I'd suggest to place such badges into the [README.md](https://github.com/tree-sitter/tree-sitter#readme) to make it easy control if something wrong [happened](https://github.com/tree-sitter/tree-sitter/issues/1216#issuecomment-871811627) with a release and to always have actual information about all components version in the well known place.

---
### TODO:
- Improvements:
  - [x] Migrate the whole `script/` folder on [xtasks](https://github.com/matklad/cargo-xtask) to make all scripts portable.
 
- CI [xtasks](https://github.com/matklad/cargo-xtask):
  - [x] Update [`pkg-config` ](https://github.com/tree-sitter/tree-sitter/issues/1608) [version](https://github.com/tree-sitter/tree-sitter/issues/1158) in Makefile.
  - [x] `cargo update` + `git commit -m "Update cargo deps"`
  - [x] Rust bindings regeneration on every C code changes.
    - [x] Use latest `bindgen` version: https://github.com/tree-sitter/tree-sitter/pull/2171
  - [ ] Check latest Emscripten version and bump its version for TS.
  - [x] Update WASM Languages for docs' playground: #1997
  - [x] Release packages on NPM:
    - [x] [`tree-sitter-cli`](https://www.npmjs.com/package/tree-sitter-cli)
    - [x] [`web-tree-sitter`](https://www.npmjs.com/package/web-tree-sitter) https://github.com/tree-sitter/tree-sitter/pull/2169 _(look to a manual task list that needs to be automated)_

- CLI [xtasks](https://github.com/matklad/cargo-xtask):
  - [x] Leverage Git for scanning TS for changed parts from the last release and inform what components require a new version.
  - [x] Automatically update deps versions for related components.
  - [ ] A task for periodically bump fixture dependency versions as part of the release process in the main repo.
         _This means to get away from using main branches all the time for fixtures._
         _**Reasons:** I saw and fixed many CI fails due to fixture changes in the last months and it's counter productive when. e.g., working on a PR  but it fails for external reasons or when merges a good PR but it unexpectedly fails on the merged commit on the master because someone merged something in some fixtures repo just several minutes ago._
         _**Fail example:** https://github.com/tree-sitter/tree-sitter-cpp/pull/202#issuecomment-1546279646_ 

- Issues:
  - [x] Check: https://github.com/tree-sitter/tree-sitter/issues/352#issuecomment-499642718
  - [x] Release latest CLI to the **crates.io**: https://github.com/tree-sitter/tree-sitter/issues/963#issuecomment-811894568
  - [x] #610 
  - [x] #820
  - [x] #1158
  - [x] #2167
  - [x] #2125
  - [x] #2237

- Extra:
  - [x] #949
  - [ ] https://github.com/tree-sitter/tree-sitter/pull/1884#issuecomment-1387393173 - check comment: https://github.com/tree-sitter/tree-sitter/issues/1880#issuecomment-1254114863

I second this, today I tried to build my library after doing cargo update and it doesn't build, because tree-sitter 0.19.2 to 0.19.5 introduces breaking changes that tree-sitter-highlight doesn't handle:

```console
error[E0107]: this struct takes 2 lifetime arguments but 1 lifetime argument was supplied
   --> /home/ryozuki/.cargo/registry/src/github.com-1ecc6299db9ec823/tree-sitter-highlight-0.19.2/src/lib.rs:104:30
    |
104 |     captures: iter::Peekable<QueryCaptures<'a, &'a [u8]>>,
    |                              ^^^^^^^^^^^^^ -- supplied 1 lifetime argument
    |                              |
    |                              expected 2 lifetime arguments
    |
note: struct defined here, with 2 lifetime parameters: `'a`, `'tree`
   --> /home/ryozuki/.cargo/registry/src/github.com-1ecc6299db9ec823/tree-sitter-0.19.5/binding_rust/lib.rs:137:12
    |
137 | pub struct QueryCaptures<'a, 'tree: 'a, T: AsRef<[u8]>> {
    |            ^^^^^^^^^^^^^ --  -----
help: add missing lifetime argument
    |
104 |     captures: iter::Peekable<QueryCaptures<'a, 'a, &'a [u8]>>,
    |                                              ^^^^

error: aborting due to previous error

For more information about this error, try `rustc --explain E0107`.
error: could not compile `tree-sitter-highlight`
```

It's 2021-08-10 and 0.20 is still not released, what's going on?

--- a/.github/workflows/emscripten.yml
+++ b/.github/workflows/emscripten.yml
@@ -1,0 +1,32 @@
+name: Update Emscripten
+
+on:
+  pull_request:
+    types: [opened, synchronize]
+
+permissions:
+  contents: write
+  pull-requests: read
+
+jobs:
+  update-emscripten:
+    if: github.actor == 'dependabot[bot]'
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v4
+        with:
+          ref: ${{ github.event.pull_request.head.sha }}
+
+      - name: Set up stable Rust toolchain
+        uses: actions-rust-lang/setup-rust-toolchain@v1
+
+      - name: Run emscripten update xtask
+        env:
+          GIT_AUTHOR_NAME: dependabot[bot]
+          GIT_AUTHOR_EMAIL: 49699333+dependabot[bot]@users.noreply.github.com
+          GIT_COMMITTER_NAME: dependabot[bot]
+          GIT_COMMITTER_EMAIL: 49699333+dependabot[bot]@users.noreply.github.com
+        run: cargo xtask update-emscripten
+
+      - name: Push updated version
+        run: git push origin HEAD:$GITHUB_HEAD_REF

--- a/xtask/Cargo.toml
+++ b/xtask/Cargo.toml
@@ -27,3 +27,4 @@ regex.workspace = true
 semver.workspace = true
 serde.workspace = true
 serde_json.workspace = true
+ureq = "2.12.1"

--- a/xtask/src/bump.rs
+++ b/xtask/src/bump.rs
@@ -1,4 +1,4 @@
-use std::{cmp::Ordering, path::Path};
+use std::cmp::Ordering;
 
 use anyhow::{anyhow, Result};
 use git2::{DiffOptions, Repository};
@@ -6,7 +6,7 @@ use indoc::indoc;
 use semver::{BuildMetadata, Prerelease, Version};
 use toml::Value;
 
-use crate::BumpVersion;
+use crate::{create_commit, BumpVersion};
 
 pub fn get_latest_tag(repo: &Repository) -> Result<String> {
     let mut tags = repo
@@ -140,42 +140,26 @@ pub fn run(args: BumpVersion) -> Result<()> {
 }
 
 fn tag_next_version(repo: &Repository, next_version: &Version) -> Result<()> {
-    // first add the manifests
-
-    let mut index = repo.index()?;
-
-    for file in [
-        "Cargo.toml",
-        "Cargo.lock",
-        "cli/Cargo.toml",
-        "cli/config/Cargo.toml",
-        "cli/loader/Cargo.toml",
-        "lib/Cargo.toml",
-        "highlight/Cargo.toml",
-        "tags/Cargo.toml",
-        "cli/npm/package.json",
-        "lib/binding_web/package.json",
-        "Makefile",
-        "lib/CMakeLists.txt",
-        "build.zig.zon",
-    ] {
-        index.add_path(Path::new(file))?;
-    }
-
-    index.write()?;
-
-    let tree_id = index.write_tree()?;
-    let tree = repo.find_tree(tree_id)?;
     let signature = repo.signature()?;
-    let parent_commit = repo.revparse_single("HEAD")?.peel_to_commit()?;
 
-    let commit_id = repo.commit(
-        Some("HEAD"),
-        &signature,
-        &signature,
+    let commit_id = create_commit(
+        repo,
         &format!("{next_version}"),
-        &tree,
-        &[&parent_commit],
+        &[
+            "Cargo.toml",
+            "Cargo.lock",
+            "cli/Cargo.toml",
+            "cli/config/Cargo.toml",
+            "cli/loader/Cargo.toml",
+            "lib/Cargo.toml",
+            "highlight/Cargo.toml",
+            "tags/Cargo.toml",
+            "cli/npm/package.json",
+            "lib/binding_web/package.json",
+            "Makefile",
+            "lib/CMakeLists.txt",
+            "build.zig.zon",
+        ],
     )?;
 
     let tag = repo.tag(

--- a/xtask/src/main.rs
+++ b/xtask/src/main.rs
@@ -5,11 +5,15 @@ mod clippy;
 mod fetch;
 mod generate;
 mod test;
+mod upgrade_emscripten;
 mod upgrade_wasmtime;
+
+use std::path::Path;
 
 use anstyle::{AnsiColor, Color, Style};
 use anyhow::Result;
 use clap::{crate_authors, Args, Command, FromArgMatches as _, Subcommand};
+use git2::{Oid, Repository};
 use semver::Version;
 
 #[derive(Subcommand)]
@@ -42,6 +46,8 @@ enum Commands {
     TestWasm,
     /// Upgrade the wasmtime dependency.
     UpgradeWasmtime(UpgradeWasmtime),
+    /// Upgrade the emscripten file.
+    UpgradeEmscripten,
 }
 
 #[derive(Args)]
@@ -208,6 +214,7 @@ fn run() -> Result<()> {
         Commands::UpgradeWasmtime(upgrade_wasmtime_options) => {
             upgrade_wasmtime::run(&upgrade_wasmtime_options)?;
         }
+        Commands::UpgradeEmscripten => upgrade_emscripten::run()?,
     }
 
     Ok(())
@@ -251,4 +258,27 @@ const fn get_styles() -> clap::builder::Styles {
                 .fg_color(Some(Color::Ansi(AnsiColor::Green))),
         )
         .placeholder(Style::new().fg_color(Some(Color::Ansi(AnsiColor::White))))
+}
+
+pub fn create_commit(repo: &Repository, msg: &str, paths: &[&str]) -> Result<Oid> {
+    let mut index = repo.index()?;
+    for path in paths {
+        index.add_path(Path::new(path))?;
+    }
+
+    index.write()?;
+
+    let tree_id = index.write_tree()?;
+    let tree = repo.find_tree(tree_id)?;
+    let signature = repo.signature()?;
+    let parent_commit = repo.revparse_single("HEAD")?.peel_to_commit()?;
+
+    Ok(repo.commit(
+        Some("HEAD"),
+        &signature,
+        &signature,
+        msg,
+        &tree,
+        &[&parent_commit],
+    )?)
 }

--- a/xtask/src/upgrade_emscripten.rs
+++ b/xtask/src/upgrade_emscripten.rs
@@ -1,0 +1,40 @@
+use std::{fs, path::Path};
+
+use anyhow::{anyhow, Result};
+use git2::Repository;
+use serde_json::Value;
+
+use crate::create_commit;
+
+pub fn run() -> Result<()> {
+    let response = ureq::get("https://api.github.com/repos/emscripten-core/emsdk/tags")
+        .call()?
+        .into_string()?;
+
+    let json = serde_json::from_str::<Value>(&response)?;
+    let version = json
+        .as_array()
+        .and_then(|arr| arr.first())
+        .and_then(|tag| tag["name"].as_str())
+        .ok_or(anyhow!("No tags found"))?;
+
+    let version_file = Path::new(env!("CARGO_MANIFEST_DIR"))
+        .parent()
+        .unwrap()
+        .join("cli")
+        .join("loader")
+        .join("emscripten-version");
+
+    fs::write(version_file, version)?;
+
+    println!("Upgraded emscripten version to {version}");
+
+    let repo = Repository::open(".")?;
+    create_commit(
+        &repo,
+        "build(deps): bump emscripten version",
+        &["cli/loader/emscripten-version"],
+    )?;
+
+    Ok(())
+}

--- a/xtask/src/upgrade_wasmtime.rs
+++ b/xtask/src/upgrade_wasmtime.rs
@@ -1,10 +1,10 @@
-use std::{path::Path, process::Command};
+use std::process::Command;
 
 use anyhow::{Context, Result};
 use git2::Repository;
 use semver::Version;
 
-use crate::UpgradeWasmtime;
+use crate::{create_commit, UpgradeWasmtime};
 
 const WASMTIME_RELEASE_URL: &str = "https://github.com/bytecodealliance/wasmtime/releases/download";
 
@@ -108,29 +108,6 @@ fn update_zig(version: &Version) -> Result<()> {
     Ok(())
 }
 
-fn create_commit(repo: &Repository, version: &Version) -> Result<()> {
-    let mut index = repo.index()?;
-    index.add_path(Path::new("lib/Cargo.toml"))?;
-    index.add_path(Path::new("build.zig.zon"))?;
-    index.write()?;
-
-    let tree_id = index.write_tree()?;
-    let tree = repo.find_tree(tree_id)?;
-    let signature = repo.signature()?;
-    let parent_commit = repo.revparse_single("HEAD")?.peel_to_commit()?;
-
-    let _ = repo.commit(
-        Some("HEAD"),
-        &signature,
-        &signature,
-        &format!("build(deps): bump wasmtime-c-api to v{version}"),
-        &tree,
-        &[&parent_commit],
-    )?;
-
-    Ok(())
-}
-
 pub fn run(args: &UpgradeWasmtime) -> Result<()> {
     println!("Upgrading wasmtime for Rust");
     update_cargo(&args.version)?;
@@ -139,7 +116,11 @@ pub fn run(args: &UpgradeWasmtime) -> Result<()> {
     update_zig(&args.version)?;
 
     let repo = Repository::open(".")?;
-    create_commit(&repo, &args.version)?;
+    create_commit(
+        &repo,
+        &format!("build(deps): bump wasmtime-c-api to v{}", args.version),
+        &["lib/Cargo.toml", "build.zig.zon"],
+    )?;
 
     Ok(())
 }

